metric/emitter/prometheus.go

```diff
@@ -0,0 +1,188 @@
+		Subsystem: "builds",
+		Name:      "finished_total",
+		Help:      "Total number of Concourse builds finished.",
+	})
+	prometheus.MustRegister(buildsFinished)
+
+	buildsSucceeded := prometheus.NewCounter(prometheus.CounterOpts{
+		Namespace: "concourse",
+		Subsystem: "builds",
+		Name:      "succeeded_total",
+		Help:      "Total number of Concourse builds succeeded.",
+	})
+	prometheus.MustRegister(buildsSucceeded)
+
+	buildsErrored := prometheus.NewCounter(prometheus.CounterOpts{
+		Namespace: "concourse",
+		Subsystem: "builds",
+		Name:      "errored_total",
+		Help:      "Total number of Concourse builds errored.",
+	})
+	prometheus.MustRegister(buildsErrored)
+
+	buildsFailed := prometheus.NewCounter(prometheus.CounterOpts{
+		Namespace: "concourse",
+		Subsystem: "builds",
+		Name:      "failed_total",
+		Help:      "Total number of Concourse builds failed.",
+	})
+	prometheus.MustRegister(buildsFailed)
+
+	buildsAborted := prometheus.NewCounter(prometheus.CounterOpts{
+		Namespace: "concourse",
+		Subsystem: "builds",
+		Name:      "aborted_total",
+		Help:      "Total number of Concourse builds aborted.",
+	})
+	prometheus.MustRegister(buildsAborted)
+
+	buildsFinishedVec := prometheus.NewCounterVec(
+		prometheus.CounterOpts{
+			Namespace: "concourse",
+			Subsystem: "builds",
+			Name:      "finished",
+			Help:      "Count of builds finished across various dimensions.",
+		},
+		[]string{"team", "pipeline", "status"},
+	)
+	prometheus.MustRegister(buildsFinishedVec)
+	buildDurationsVec := prometheus.NewHistogramVec(
+		prometheus.HistogramOpts{
+			Namespace: "concourse",
+			Subsystem: "builds",
+			Name:      "duration_seconds",
+			Help:      "Build time in seconds",
+		},
+		[]string{"team", "pipeline"},
+	)
+	prometheus.MustRegister(buildDurationsVec)
+
+	go func() {
+		http.Handle("/metrics", promhttp.Handler())
+		// It would be nice to be able to log here, or maybe we just panic stop?
+		// err := http.ListenAndServe(config.Address, nil)
+		http.ListenAndServe(config.Address, nil)
+		//if err != nil {
+		//	logger.Error("failed-to-initiate-prometheus-endpoint", err)
+		//}
+	}()
+
+	return &PrometheusEmitter{
+		buildsStarted:     buildsStarted,
+		buildsFinished:    buildsFinished,
+		buildsFinishedVec: buildsFinishedVec,
+		buildDurationsVec: buildDurationsVec,
+		buildsSucceeded:   buildsSucceeded,
+		buildsErrored:     buildsErrored,
+		buildsFailed:      buildsFailed,
+		buildsAborted:     buildsAborted,
+	}, nil
+}
+
+func (emitter *PrometheusEmitter) Emit(logger lager.Logger, event metric.Event) {
+	// In order to provide idiomatic Prometheus metrics, we'll have to convert the various
+	// Event types (differentiated by the less-than-ideal string Name field) into different
+	// Prometheus metrics.
+	switch event.Name {
+	case "build started":
+		emitter.buildsStarted.Inc()
+	case "build finished":
+		emitter.buildFinishedMetrics(logger, event)
+	default:
+		// unless we have a specific metric, we do nothing
```

Comment:
Hmm I think it's probably worth fleshing out the rest. I understand that the initial use case is just for tracking build stats, but there are plenty of other metrics that are simple gauges that should work well with this, no?

Reply:
Yeah, absolutely. I was just doing the ones I was really interested in because I was constrained by time. But I can flesh out the ones that are gauges, sure.

Reply:
Also happy to merge as-is and accept those as subsequent PRs, just wanted to make sure our goals are aligned. :)